torch/csrc/autograd/python_variable.cpp

```diff
@@ -485,6 +489,22 @@ namespace torch { namespace autograd {
 extern PyMethodDef variable_methods[];
 extern void initTorchFunctions(PyObject *module);
 
+void initTensorImplConversion(PyObject* module) {
+  auto m = py::handle(module).cast<py::module>();
+  m.def("_wrap_tensor_impl", [](void* ptr) {
+    auto p =
+        c10::intrusive_ptr<c10::TensorImpl, at::UndefinedTensorImpl>::reclaim(
```

Comment:
Use of `reclaim` was unexpected to me. This means that if you call `_tensor_impl_raw_handle` and then don't do anything with the result, you'll leak the tensor. The method name doesn't suggest anything like that!

I think the root cause of the problem is there isn't any API in `intrusive_ptr` which says "given a non-owning `T*`, mint a new owning reference to it, increasing the refcount." An API like that would be the most appropriate thing to use.  Maybe something like:

```
  template <typename T>
  inline c10::intrusive_ptr<T> intrusive_from_nonowning(T* self) {
    auto ptr = c10::intrusive_ptr<intrusive_ptr_target>::reclaim(self);
    auto ptr_copy = ptr;
    ptr_copy.release();
    return ptr;
  }
```

It's a bit naughty use of `reclaim` but if you inline all the implementations, the sequence of increfs/decrefs is correct (one incref from the copy, no decrefs).

Reply:
cc @smessmer